beacon-chain/sync/backfill/service.go

```diff
@@ -0,0 +1,277 @@
+const defaultBatchSize = 64
+
+type Service struct {
+	ctx           context.Context
+	su            *StatusUpdater
+	ms            minimumSlotter
+	cw            startup.ClockWaiter
+	nWorkers      int
+	errChan       chan error
+	batchSeq      *batchSequencer
+	batchSize     uint64
+	pool          BatchWorkerPool
+	verifier      *verifier
+	p2p           p2p.P2P
+	batchImporter batchImporter
+}
+
+var _ runtime.Service = (*Service)(nil)
+
+type ServiceOption func(*Service) error
+
+func WithWorkerCount(n int) ServiceOption {
+	return func(s *Service) error {
+		s.nWorkers = n
+		return nil
+	}
+}
+
+func WithBatchSize(n uint64) ServiceOption {
+	return func(s *Service) error {
+		s.batchSize = n
+		return nil
+	}
+}
+
+type minimumSlotter interface {
+	minimumSlot() primitives.Slot
+	setClock(*startup.Clock)
+}
+
+type defaultMinimumSlotter struct {
+	clock *startup.Clock
+	cw    startup.ClockWaiter
+	ctx   context.Context
+}
+
+func (d defaultMinimumSlotter) minimumSlot() primitives.Slot {
+	if d.clock == nil {
+		var err error
+		d.clock, err = d.cw.WaitForClock(d.ctx)
+		if err != nil {
+			log.WithError(err).Fatal("failed to obtain system/genesis clock, unable to start backfill service")
+		}
+	}
+	return MinimumBackfillSlot(d.clock.CurrentSlot())
+}
+
+func (d defaultMinimumSlotter) setClock(c *startup.Clock) {
+	d.clock = c
+}
+
+var _ minimumSlotter = &defaultMinimumSlotter{}
+
+type batchImporter func(ctx context.Context, b batch, su *StatusUpdater) (*dbval.BackfillStatus, error)
+
+func defaultBatchImporter(ctx context.Context, b batch, su *StatusUpdater) (*dbval.BackfillStatus, error) {
+	status := su.status()
+	if err := b.ensureParent(bytesutil.ToBytes32(status.LowParentRoot)); err != nil {
+		return status, err
+	}
+	// Import blocks to db and update db state to reflect the newly imported blocks.
+	// Other parts of the beacon node may use the same StatusUpdater instance
+	// via the coverage.AvailableBlocker interface to safely determine if a given slot has been backfilled.
+	status, err := su.fillBack(ctx, b.results)
+	if err != nil {
+		log.WithError(err).Fatal("Non-recoverable db error in backfill service, quitting.")
+	}
+	return status, nil
+}
+
+func NewService(ctx context.Context, su *StatusUpdater, cw startup.ClockWaiter, p p2p.P2P, opts ...ServiceOption) (*Service, error) {
+	s := &Service{
+		ctx:           ctx,
+		su:            su,
+		cw:            cw,
+		ms:            &defaultMinimumSlotter{cw: cw, ctx: ctx},
+		p2p:           p,
+		batchImporter: defaultBatchImporter,
+	}
+	for _, o := range opts {
+		if err := o(s); err != nil {
+			return nil, err
+		}
+	}
+	if s.nWorkers == 0 {
+		s.nWorkers = defaultWorkerCount
+	}
+	if s.batchSize == 0 {
+		s.batchSize = defaultBatchSize
+	}
+	s.pool = newP2PBatchWorkerPool(p, s.nWorkers)
+
+	return s, nil
+}
+
+func (s *Service) initVerifier(ctx context.Context) (*verifier, error) {
+	cps, err := s.su.originState(ctx)
```

Comment:
Could there be a scenario where the to-be-validated blocks are newer than the origin state?

Reply:
No, I don't think so. The origin state is the state used to initiate the checkpoint sync, so it is by definition the state we are backfilling from, and it must be finalized, and the descendant of all the blocks that we backfill. 

A future case to consider, which we don't currently support, is triggering a new checkpoint sync on top of an existing db - see https://github.com/prysmaticlabs/prysm/issues/13003

Once we do enable that feature, we should always write the updated origin to the db and use that state for the validator registry. Note that we may need to track additional data in the status db record to support this, because currently we only track the low/high end of one range of blocks to backfill, and stategen uses the low end in this db record to figure out if blocks are available. if we allow multiple checkpoint syncs against the same db, we need to track the target of the backfill separately from the lowest available slot, and we need to account for "holes" in the range.